File: 79WordSearch.go
package main

import (
	"fmt"
)

func exist(board [][]byte, word string) bool {
	reachedGrid := make([][]bool, len(board))
	for index := range reachedGrid {
		reachedGrid[index] = make([]bool, len(board[0]))
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			//find the word
			var funcFindWord func(int, int, int) bool
			funcFindWord = func(x, y, startIndexOfSubString int) bool {
				if x >= 0 && x < len(board) && y >= 0 && y < len(board[x]) &&
					!reachedGrid[x][y] && board[x][y] == word[startIndexOfSubString] {
					//fmt.Println(x, y, board[x][y], startIndexOfSubString, word[startIndexOfSubString])
					reachedGrid[x][y] = true

					if startIndexOfSubString == len(word)-1 {
						return true
					}

					// check all neighbor grids
					if funcFindWord(x-1, y, startIndexOfSubString+1) ||
						funcFindWord(x+1, y, startIndexOfSubString+1) ||
						funcFindWord(x, y-1, startIndexOfSubString+1) ||
						funcFindWord(x, y+1, startIndexOfSubString+1) {
						return true
					} else {
						reachedGrid[x][y] = false
						return false
					}
				}

				return false
			}

			if funcFindWord(i, j, 0) {
				return true
			}
		}
	}

	return false
}

func main() {
	output := exist([][]byte{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}}, "ABCCED")
	output1 := exist([][]byte{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}}, "SEE")
	output2 := exist([][]byte{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}}, "ABCB")

	fmt.Println("\n******** output *********\n", output, output1, output2)
}

File: cpp/79WordSearch.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <cstdint>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

template<class T>
void printMap(const T& container) {
//    return;

    cout << "size " << container.size() << endl;
    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[ ";
        }

        cout << "[" << it->first << ", " << it->second << "]";

        if (next(it) == container.end()) {
            cout << " ]" << endl;
        } else {
            cout << ", ";
        }
    }
}


/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void printList(const ListNode* node) {
    while (node != nullptr) {
        cout << node->val;

        node = node->next;
        
        if (node != nullptr) {
            cout << "->";
        } else {
            cout << endl;
        }
    }
}

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {

        function<int(int,int)> convert = [](int x, int y) {
            return x * 1000 + y;
        };
        function<int(int)> getX = [](int index) {
            return index / 1000;
        };
        function<int(int)> getY = [](int index) {
            return index % 1000;
        };

        unordered_map<char, vector<int>> mapCharIndex;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j< board[0].size(); j++) {
                auto it = mapCharIndex.find(board[i][j]);
                if (it != mapCharIndex.end()) {
                    it->second.push_back(convert(i, j));
                } else {
                    mapCharIndex.emplace(board[i][j], vector<int>{convert(i, j)});
                }
            }
        }

        unordered_set<int> visited;

        function<bool(int, int)> findPath =
            [&findPath, &visited, &mapCharIndex, &board, &word, &convert, &getX, &getY](int curBoardIndex, int curCharIndex) {

                char curChar = word[curCharIndex];
                int x = getX(curBoardIndex);
                int y = getY(curBoardIndex);

                if (curChar != board[x][y]) {
                    return false;
                }

                if (visited.find(curBoardIndex) != visited.end()) {
                    return false;
                }

                int nextCharIndex = curCharIndex + 1;
                if (nextCharIndex == word.size()) {
                    return true;
                }

                visited.emplace(curBoardIndex);

                if (x - 1 >= 0 && findPath(convert(x-1, y), nextCharIndex)) {
                    return true;
                }
                if (y - 1 >= 0 && findPath(convert(x, y-1), nextCharIndex)) {
                    return true;
                }
                if (x + 1 < board.size() && findPath(convert(x+1, y), nextCharIndex)) {
                    return true;
                }
                if (y + 1 < board[0].size() && findPath(convert(x, y+1), nextCharIndex)) {
                    return true;
                }

                visited.erase(curBoardIndex);

                return false;
            };

        for (const auto& index : mapCharIndex[word[0]]) {
            if (findPath(index, 0)) {
                return true;
            }
        }

        return false;
    }
};

int main(){
    Solution s;

    vector<vector<char>> input{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
	auto output = s.exist(input, "ABCCED");

    vector<vector<char>> input1{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
	auto output1 = s.exist(input1, "SEE");

    vector<vector<char>> input2{{'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
	auto output2 = s.exist(input2, "ABCB");

    vector<vector<char>> input3{{'a'}};
	auto output3 = s.exist(input3, "a");

    cout << "*************output*************" << endl;
    cout << output << endl;
    cout << output1 << endl;
    cout << output2 << endl;
    cout << output3 << endl;

    return 0;
}
